kv/kvserver: TestLeaseQueueSwitchesLeaseType failed

kv/kvserver.TestLeaseQueueSwitchesLeaseType [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17346234?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17346234?buildTab=artifacts#/) on master @ [88faab92fdf2566d3ccb710374aa9f285f077906](https://github.com/cockroachdb/cockroach/commits/88faab92fdf2566d3ccb710374aa9f285f077906):


```
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_arm64.s:1222
* Wraps: (2) panic: can only defortify at current term if told by the leader or if fortification has expired
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
panic: can only defortify at current term if told by the leader or if fortification has expired [recovered]
	panic: can only defortify at current term if told by the leader or if fortification has expired

goroutine 77414 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x400c1e9218?, {0x8476ed0, 0x40073907b0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:226 +0x68
panic({0x5cd5b20?, 0x400c02a590?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/raft.assertTrue(...)
	github.com/cockroachdb/cockroach/pkg/raft/util.go:326
github.com/cockroachdb/cockroach/pkg/raft.(*raft).deFortify(0x2?, 0x400574dd80?, 0x17ff0ef5c34e3075?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:2310 +0xc8
github.com/cockroachdb/cockroach/pkg/raft.(*raft).sendDeFortify(0x4002ecd6b0?, 0x400d273200?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:838 +0x38
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify.func1(0x60dc5a0?, 0x4002ecd6b0?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:887 +0x28
github.com/cockroachdb/cockroach/pkg/raft/tracker.(*ProgressTracker).Visit(0x40092da958, 0x400c1e9518)
	github.com/cockroachdb/cockroach/pkg/raft/tracker/progresstracker.go:106 +0x180
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify(0x40092da900)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:886 +0x5c
github.com/cockroachdb/cockroach/pkg/raft.(*raft).tickElection(0x40092da900)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:1078 +0x58
github.com/cockroachdb/cockroach/pkg/raft.(*RawNode).Tick(...)
	github.com/cockroachdb/cockroach/pkg/raft/rawnode.go:68
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick(0x400d64a388, {0x8476ed0, 0x4005964690}, 0x401123e720, 0x401a117b48)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515 +0x7c8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick(0x400b941c08, {0x4008576e10?, 0x1?}, 0x1?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711 +0x140
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0x400af2c770, {0x8476ed0, 0x40073907b0}, {0x849cf20, 0x400b941c08}, 0x4006329208)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410 +0x3b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x8476ed0?, 0x40073907b0?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319 +0x4c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x6ac3442?, 0x4007b73a40?})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498 +0x198
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 77337
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x384
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueSwitchesLeaseType.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-43282

kv/kvserver.TestLeaseQueueSwitchesLeaseType [failed](https://tanzanite.cluster.engflow.com/invocations/default/b95364b6-14d1-487b-9393-e48237da7d11?testReportRun=12&testReportShard=36&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [42f40f59cae3c0fd8842e194d6991c951ab4382f](https://github.com/cockroachdb/cockroach/commits/42f40f59cae3c0fd8842e194d6991c951ab4382f):

Fatal error:

```
panic: can only defortify at current term if told by the leader or if fortification has expired [recovered]
	panic: can only defortify at current term if told by the leader or if fortification has expired

```
Stack: 

```
goroutine 66922 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc015987248?, {0x8d93c38, 0xc004e077a0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:226 +0x65
panic({0x65ec4e0?, 0xc00bb8f520?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/raft.assertTrue(...)
	github.com/cockroachdb/cockroach/pkg/raft/util.go:326
github.com/cockroachdb/cockroach/pkg/raft.(*raft).deFortify(0x1a?, 0x1?, 0x2?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:2310 +0xa5
github.com/cockroachdb/cockroach/pkg/raft.(*raft).sendDeFortify(0xc00ae10510?, 0xc00922d3b0?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:838 +0x35
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify.func1(0x69f3460?, 0xc00ae10510?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:887 +0x1d
github.com/cockroachdb/cockroach/pkg/raft/tracker.(*ProgressTracker).Visit(0xc00c09cc58, 0xc015987540)
	github.com/cockroachdb/cockroach/pkg/raft/tracker/progresstracker.go:106 +0x1d5
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify(0xc00c09cc00)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:886 +0x5e
github.com/cockroachdb/cockroach/pkg/raft.(*raft).tickElection(0xc00c09cc00)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:1078 +0x6b
github.com/cockroachdb/cockroach/pkg/raft.(*RawNode).Tick(...)
	github.com/cockroachdb/cockroach/pkg/raft/rawnode.go:68
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick(0xc0028c8388, {0x8d93c38, 0xc00d92c090}, 0xc00827f7a0, 0xc013156000)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515 +0xa0e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick(0xc00754d008, {0xc0047608d0?, 0x2?}, 0x2?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711 +0x187
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc00e8ed3b0, {0x8d93c38, 0xc004e077a0}, {0x8db9ae0, 0xc00754d008}, 0xc006142008)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410 +0x395
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x8d93c38?, 0xc004e077a0?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x73da6b2?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498 +0x1f0
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 66813
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x49c
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLeaseQueueSwitchesLeaseType
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestLeaseQueueSwitchesLeaseType2028716348
    test_log_scope.go:76: use -show-logs to present logs inline
    lease_queue_test.go:305: initial: epochLeases=324 leaderLeases=0 expLeases=2
    lease_queue_test.go:310: 100%: epochLeases=6 leaderLeases=303 expLeases=2
    lease_queue_test.go:310: 100%: epochLeases=1 leaderLeases=323 expLeases=2
    lease_queue_test.go:310: 100%: epochLeases=1 leaderLeases=323 expLeases=2
    lease_queue_test.go:310: 100%: epochLeases=1 leaderLeases=323 expLeases=2
    lease_queue_test.go:310: 100%: epochLeases=0 leaderLeases=324 expLeases=2
*
* ERROR: a panic has occurred!
* panic: can only defortify at current term if told by the leader or if fortification has expired
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/raft.assertTrue
*   | 	github.com/cockroachdb/cockroach/pkg/raft/util.go:326
*   | github.com/cockroachdb/cockroach/pkg/raft.(*raft).deFortify
*   | 	github.com/cockroachdb/cockroach/pkg/raft/raft.go:2310
*   | github.com/cockroachdb/cockroach/pkg/raft.(*raft).sendDeFortify
*   | 	github.com/cockroachdb/cockroach/pkg/raft/raft.go:838
*   | github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify.func1
*   | 	github.com/cockroachdb/cockroach/pkg/raft/raft.go:887
*   | github.com/cockroachdb/cockroach/pkg/raft/tracker.(*ProgressTracker).Visit
*   | 	github.com/cockroachdb/cockroach/pkg/raft/tracker/progresstracker.go:106
*   | github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify
*   | 	github.com/cockroachdb/cockroach/pkg/raft/raft.go:886
*   | github.com/cockroachdb/cockroach/pkg/raft.(*raft).tickElection
*   | 	github.com/cockroachdb/cockroach/pkg/raft/raft.go:1078
*   | github.com/cockroachdb/cockroach/pkg/raft.(*RawNode).Tick
*   | 	github.com/cockroachdb/cockroach/pkg/raft/rawnode.go:68
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (2) panic: can only defortify at current term if told by the leader or if fortification has expired
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=36</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueSwitchesLeaseType.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

kv/kvserver.TestLeaseQueueSwitchesLeaseType [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17355189?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17355189?buildTab=artifacts#/) on master @ [3fe7e6af81446821dcb8e2146d3b8c41622a4231](https://github.com/cockroachdb/cockroach/commits/3fe7e6af81446821dcb8e2146d3b8c41622a4231):


```
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_arm64.s:1222
* Wraps: (2) panic: can only defortify at current term if told by the leader or if fortification has expired
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
panic: can only defortify at current term if told by the leader or if fortification has expired [recovered]
	panic: can only defortify at current term if told by the leader or if fortification has expired

goroutine 75695 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x401b139218?, {0x847a350, 0x4009813bf0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:226 +0x68
panic({0x5cd7d80?, 0x400a70fef0?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/raft.assertTrue(...)
	github.com/cockroachdb/cockroach/pkg/raft/util.go:326
github.com/cockroachdb/cockroach/pkg/raft.(*raft).deFortify(0x3?, 0x0?, 0x17ff46698cff25b3?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:2310 +0xc8
github.com/cockroachdb/cockroach/pkg/raft.(*raft).sendDeFortify(0x4011022720?, 0x4003319290?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:838 +0x38
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify.func1(0x60dea00?, 0x4011022720?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:887 +0x28
github.com/cockroachdb/cockroach/pkg/raft/tracker.(*ProgressTracker).Visit(0x400724bcd8, 0x401b139518)
	github.com/cockroachdb/cockroach/pkg/raft/tracker/progresstracker.go:106 +0x180
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify(0x400724bc80)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:886 +0x5c
github.com/cockroachdb/cockroach/pkg/raft.(*raft).tickElection(0x400724bc80)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:1078 +0x58
github.com/cockroachdb/cockroach/pkg/raft.(*RawNode).Tick(...)
	github.com/cockroachdb/cockroach/pkg/raft/rawnode.go:68
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick(0x4011178d88, {0x847a350, 0x4013c6ff50}, 0x40112b8e70, 0x40037a2b58)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515 +0x7c8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick(0x400b2da008, {0x400a898060?, 0xac?}, 0x124?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711 +0x140
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0x40037bec40, {0x847a350, 0x4009813bf0}, {0x84a03a0, 0x400b2da008}, 0x4004ff4c88)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410 +0x3b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x847a350?, 0x4009813bf0?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319 +0x4c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x6ac5c62?, 0x400fdcc798?})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498 +0x198
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 75530
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x384
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueSwitchesLeaseType.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

kv/kvserver.TestLeaseQueueSwitchesLeaseType [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17355663?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17355663?buildTab=artifacts#/) on master @ [0ae057361aa19196db1385f6617eb82ad22340a9](https://github.com/cockroachdb/cockroach/commits/0ae057361aa19196db1385f6617eb82ad22340a9):


```
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_arm64.s:1222
* Wraps: (2) panic: can only defortify at current term if told by the leader or if fortification has expired
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
panic: can only defortify at current term if told by the leader or if fortification has expired [recovered]
	panic: can only defortify at current term if told by the leader or if fortification has expired

goroutine 80715 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x40137ef218?, {0x847a270, 0x4012e31410})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:226 +0x68
panic({0x5cd7ce0?, 0x400a91ca00?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/raft.assertTrue(...)
	github.com/cockroachdb/cockroach/pkg/raft/util.go:326
github.com/cockroachdb/cockroach/pkg/raft.(*raft).deFortify(0x1a?, 0x2?, 0x3?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:2310 +0xc8
github.com/cockroachdb/cockroach/pkg/raft.(*raft).sendDeFortify(0x40184e6de0?, 0x40110fb710?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:838 +0x38
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify.func1(0x60de960?, 0x40184e6de0?)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:887 +0x28
github.com/cockroachdb/cockroach/pkg/raft/tracker.(*ProgressTracker).Visit(0x4006dd0ad8, 0x40137ef518)
	github.com/cockroachdb/cockroach/pkg/raft/tracker/progresstracker.go:106 +0x180
github.com/cockroachdb/cockroach/pkg/raft.(*raft).bcastDeFortify(0x4006dd0a80)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:886 +0x5c
github.com/cockroachdb/cockroach/pkg/raft.(*raft).tickElection(0x4006dd0a80)
	github.com/cockroachdb/cockroach/pkg/raft/raft.go:1078 +0x58
github.com/cockroachdb/cockroach/pkg/raft.(*RawNode).Tick(...)
	github.com/cockroachdb/cockroach/pkg/raft/rawnode.go:68
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick(0x4014878008, {0x847a270, 0x4005ec0150}, 0x4019e1a210, 0x401655c7b0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1515 +0x7c8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick(0x400c4e6008, {0x29496?, 0x2?}, 0x2?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:711 +0x140
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0x4010a3aaf0, {0x847a270, 0x4012e31410}, {0x84a02c0, 0x400c4e6008}, 0x4007aa2c88)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:410 +0x3b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x847a270?, 0x4012e31410?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:319 +0x4c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x6ac5bc2?, 0x400e80af98?})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:498 +0x198
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 80557
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x384
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueSwitchesLeaseType.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/raft/raft.go
+++ b/pkg/raft/raft.go
@@ -832,13 +832,27 @@ func (r *raft) sendFortify(to pb.PeerID) {
 
 // sendDeFortify sends a de-fortification RPC to the given peer.
 func (r *raft) sendDeFortify(to pb.PeerID) {
+	fortifiedTerm := r.fortificationTracker.Term()
 	if to == r.id {
 		// We handle the case where the leader is trying to de-fortify itself
 		// specially. Doing so avoids a self-addressed message.
-		r.deFortify(r.id, r.fortificationTracker.Term())
+		switch {
+		case r.Term == fortifiedTerm:
+			r.deFortify(r.id, fortifiedTerm)
+		case r.Term > fortifiedTerm:
+			// NB: The current term has advanced, so we don't attempt to de-fortify, as the de-fortification
+			// attempt corresponds to a prior term. These term checks would have happened in Step had we
+			// sent a self-addressed message; we decided not to, so we need to handle this especially
+			// ourselves.
+			r.logger.Debugf("de-foritfying self at term %d is a no-op; current term %d",
+				fortifiedTerm, r.Term,
+			)
+		case r.Term < fortifiedTerm:
+			panic("fortification tracker's term cannot be higher than raft group's")
+		}
 		return
 	}
-	r.send(pb.Message{To: to, Type: pb.MsgDeFortifyLeader, Term: r.fortificationTracker.Term()})
+	r.send(pb.Message{To: to, Type: pb.MsgDeFortifyLeader, Term: fortifiedTerm})
 }
 
 // bcastAppend sends RPC, with entries to all peers that are not up-to-date
@@ -2305,6 +2319,21 @@ func (r *raft) handleDeFortify(m pb.Message) {
 }
 
 // deFortify revokes previously provided fortification to a leader.
+//
+// The from argument corresponds to the ID of the peer initiating (directly[1]
+// or indirectly[2]) the de-fortification. The term argument corresponds to the
+// term known to the peer[3] initiating de-fortification.
+//
+// [1] For example, if a peer realizes it's no longer supporting a previously
+// fortified leader or if a previously fortified leader steps down and
+// broadcasts MsgDeFortifyLeader to the raft group.
+//
+// [2] For example, if a new leader is elected without a vote from this peer
+// (who continues to fortify the previous leader). The peer then learns about
+// the previous leadership term ending when the new leader contacts it.
+//
+// [3] Note that this may not be the term being de-fortified. The term being
+// de-fortified is r.Term, which is the term of the peer being de-fortified.
 func (r *raft) deFortify(from pb.PeerID, term uint64) {
 	assertTrue(
 		// We're not currently fortified, so de-fortification is a no-op...

